File: Javascript/1493. Longest Subarray of 1's After Deleting One Element.js
// Given a binary array nums, you should delete one element from it.

// Return the size of the longest non-empty subarray containing only 1's in the resulting array. Return 0 if there is no such subarray.

 

// Example 1:

// Input: nums = [1,1,0,1]
// Output: 3
// Explanation: After deleting the number in position 2, [1,1,1] contains 3 numbers with value of 1's.
// Example 2:

// Input: nums = [0,1,1,1,0,1,1,0,1]
// Output: 5
// Explanation: After deleting the number in position 4, [0,1,1,1,1,1,0,1] longest subarray with value of 1's is [1,1,1,1,1].
// Example 3:

// Input: nums = [1,1,1]
// Output: 2
// Explanation: You must delete one element.

/**
 * @param {number[]} nums
 * @return {number}
 */
var longestSubarray = function(nums) {
    let map = {0: 0, 1: 0}
    let i = 0
    let j = 0
    let longest = Number.MIN_SAFE_INTEGER
    while(j < nums.length){
        map[nums[j].toString()] += 1
        // Sliding Window
        while(map['0'] > 1){
            map[nums[i].toString()] -= 1
            i += 1
        }
        if (map['0'] == 0){
            // No element 0, must delete one element 1
            longest = Math.max(longest, map['1'] - 1) 
        }else{
            longest = Math.max(longest, map['1'])
        }
        j += 1
    }
    return longest === Number.MIN_SAFE_INTEGER ? 0: longest
};

File: Python/1493. Longest Subarray of 1's After Deleting One Element.py
# Given a binary array nums, you should delete one element from it.

# Return the size of the longest non-empty subarray containing only 1's in the resulting array. Return 0 if there is no such subarray.

 

# Example 1:

# Input: nums = [1,1,0,1]
# Output: 3
# Explanation: After deleting the number in position 2, [1,1,1] contains 3 numbers with value of 1's.
# Example 2:

# Input: nums = [0,1,1,1,0,1,1,0,1]
# Output: 5
# Explanation: After deleting the number in position 4, [0,1,1,1,1,1,0,1] longest subarray with value of 1's is [1,1,1,1,1].
# Example 3:

# Input: nums = [1,1,1]
# Output: 2
# Explanation: You must delete one element.

class Solution(object):
    def longestSubarray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        zero_count = 0
        left, right = 0, 0
        n = len(nums)
        max_length = 0        
        for right in range(0, n):
            if nums[right] == 0:
                zero_count += 1
            while zero_count > 1:
                if nums[left] == 0:
                    zero_count -= 1
                left += 1
            curr_length = right - left # Including curr second zeros
            max_length = max(max_length, curr_length) # curr_length - curr zero
        
        return max_length
            